File: app/src/main/java/com/agrogames/islandsofwar/engine/impl/bullet/Bullet.java
package com.agrogames.islandsofwar.engine.impl.bullet;

import android.os.Build;

import androidx.annotation.RequiresApi;

import com.agrogames.islandsofwar.engine.abs.graphics.GraphicsAdder;
import com.agrogames.islandsofwar.engine.abs.bullet.IBullet;
import com.agrogames.islandsofwar.engine.abs.renderable.RenderableObject;
import com.agrogames.islandsofwar.factories.Factory;
import com.agrogames.islandsofwar.map.impl.Water;
import com.agrogames.islandsofwar.engine.abs.common.Cell;
import com.agrogames.islandsofwar.engine.abs.common.Point;
import com.agrogames.islandsofwar.engine.abs.bullet.BulletAdder;
import com.agrogames.islandsofwar.engine.abs.map.MapObject;
import com.agrogames.islandsofwar.engine.abs.unit.IUnit;
import com.agrogames.islandsofwar.engine.abs.map.MapProvider;
import com.agrogames.islandsofwar.engine.abs.unit.IUnitAdder;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Bullet implements IBullet {
    private final String texture;
    private final String creationSound;
    private Point location;
    private float rotation;
    private final float speed;
    private final int power;
    private final float longRange;
    private final int flightHeight;
    private final int targetHeight;
    private final IUnit owner;
    private final boolean bang;
    private final float bangRange;
    private boolean isFlying = true;
    private float flewDistance;
    private Cell goal;

    public Bullet(String texture, String creationSound, Point location, float speed, int power, float longRange, int flightHeight, int targetHeight, IUnit owner, boolean bang, float bangRange) {
        this.creationSound = creationSound;
        this.longRange = longRange;
        this.texture = texture;
        this.location = location;
        this.speed = speed;
        this.power = power;
        this.flightHeight = flightHeight;
        this.targetHeight = targetHeight;
        this.owner = owner;
        this.bang = bang;
        this.bangRange = bangRange;
    }

    @Override
    public String getTexture() {
        return texture;
    }

    @Override
    public boolean hasStopped() {
        return !isFlying;
    }

    @Override
    public Point getLocation() {
        return location;
    }

    @Override
    public float getRotation() {
        return rotation;
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    public void update(MapProvider provider, BulletAdder bulletAdder, IUnitAdder unitAdder, GraphicsAdder graphicsAdder, float deltaTime) {
        move(deltaTime);
        MapObject[] all = provider.getAll();
        MapObject[] enemies = enemyAt(new Cell((int) location.x + 1, (int) location.y + 1), all, flightHeight);
        if(enemies.length != 0){
            for(MapObject enemy : enemies){
                if(enemy instanceof IUnit) ((IUnit)enemy).loseHealth(power);
                stop(provider, graphicsAdder);
                graphicsAdder.addGraphics(Factory.getGraphics("hit", getLocation(), getRotation()));
            }
        }
        else if(new Cell((int) location.x + 1, (int) location.y + 1).equals(goal) || flewDistance >= longRange){
            MapObject[] e2 = enemyAt(new Cell((int) location.x + 1, (int) location.y + 1), all, targetHeight);
            for(MapObject enemy : e2){
                if(enemy instanceof IUnit) ((IUnit)enemy).loseHealth(power);
                graphicsAdder.addGraphics(Factory.getGraphics("hit", getLocation(), getRotation()));
            }
            stop(provider, graphicsAdder);
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private void stop(MapProvider provider, GraphicsAdder adder){
        isFlying = false;
        if(bang && Arrays.stream(provider.getAll()).noneMatch(o ->
                Arrays.asList(o.getTerritory()).contains(new Cell(location)) && o instanceof Water) && targetHeight == 1){
            adder.addGraphics(Factory.getGraphics("bang", getLocation(), getRotation()));
        }
    }

    private void move(float deltaTime){
        double dx = Math.cos(rotation) * (double) speed * (double) deltaTime;
        double dy = Math.sin(rotation) * (double) speed * (double) deltaTime;
        location = new Point(location.x + (float)dx,  location.y + (float) dy);
        flewDistance += speed * deltaTime;
    }

    private MapObject[] enemyAt(Cell cell, MapObject[] enemies, int height){
        if (bang) {
            List<MapObject> e = new ArrayList<>();
            for (MapObject object : enemies){
                if(object != owner && object.getHeight() == height){
                    Cell[] territory = object.getTerritory();
                    for (Cell c : territory){
                        if(cell.equals(c)){
                            if(!e.contains(object)) e.add(object);
                        }
                    }
                    if(object instanceof RenderableObject){
                        Point l = ((RenderableObject) object).getLocation();
                        if(l.x - bangRange < location.x && l.x + bangRange > location.x &&
                                l.y - bangRange < location.y && l.y + bangRange > location.y){
                            if(!e.contains(object)) e.add(object);
                        }
                    }
                }
            }
            return e.toArray(new MapObject[0]);
        } else {
            for (MapObject object : enemies){
                if(object != owner && object.getHeight() == height){
                    Cell[] territory = object.getTerritory();
                    for (Cell c : territory){
                        if(cell.equals(c)){
                            return new MapObject[]{object};
                        }
                    }
                }
            }
        }
        return new MapObject[0];
    }

    @Override
    public void setGoal(Point goal) {
        this.rotation = (float) Math.atan(
                ((double) goal.y - (double) location.y) /
                        ((double) goal.x - (double) location.x));
        if (goal.x  < location.x){
            rotation += Math.PI;
        }
        this.goal = new Cell(goal);
    }

    @Override
    public String getCreationSound() {
        return creationSound;
    }
}

File: app/src/main/java/ru/agrogames/islands/engine/impl/bullet/Bullet.kt
package ru.agrogames.islands.engine.impl.bullet

import ru.agrogames.islands.engine.abs.bullet.BulletAdder
import ru.agrogames.islands.engine.abs.bullet.IBullet
import ru.agrogames.islands.engine.abs.common.Cell
import ru.agrogames.islands.engine.abs.common.Point
import ru.agrogames.islands.engine.abs.graphics.GraphicsAdder
import ru.agrogames.islands.engine.abs.map.MapObject
import ru.agrogames.islands.engine.abs.map.MapProvider
import ru.agrogames.islands.engine.abs.renderable.RenderableObject
import ru.agrogames.islands.engine.abs.unit.IUnit
import ru.agrogames.islands.engine.abs.unit.IUnitAdder
import ru.agrogames.islands.factories.Factory
import ru.agrogames.islands.map.Water

class Bullet(override val texture: String, override val creationSound: String?, override var location: Point, private val speed: Float, private val power: Int, private val longRange: Float, private val flightHeight: Int, private val targetHeight: Int, private val owner: IUnit?, private val bang: Boolean, private val bangRange: Float) : IBullet {
    override var rotation = 0f
        private set
    private var isFlying = true
    private var flewDistance = 0f
    private var goal: Cell? = null
    override fun hasStopped(): Boolean {
        return !isFlying
    }

    override fun update(provider: MapProvider, bulletAdder: BulletAdder, unitAdder: IUnitAdder, graphicsAdder: GraphicsAdder, deltaTime: Float) {
        move(deltaTime)
        val all = provider.all
        val enemies = enemyAt(Cell(location.x.toInt() + 1, location.y.toInt() + 1), all, flightHeight)
        if (enemies.size != 0) {
            for (enemy in enemies) {
                if (enemy is IUnit) enemy.loseHealth(power)
                stop(provider, graphicsAdder)
                graphicsAdder.addGraphics(Factory.getGraphics("hit", location, rotation))
            }
        } else if (Cell(location.x.toInt() + 1, location.y.toInt() + 1) == goal || flewDistance >= longRange) {
            val e2 = enemyAt(Cell(location.x.toInt() + 1, location.y.toInt() + 1), all, targetHeight)
            for (enemy in e2) {
                if (enemy is IUnit) enemy.loseHealth(power)
                graphicsAdder.addGraphics(Factory.getGraphics("hit", location, rotation))
            }
            stop(provider, graphicsAdder)
        }
    }

    private fun stop(provider: MapProvider, adder: GraphicsAdder) {
        isFlying = false
        if (bang && provider.all.none { it.territory.contains(Cell(location)) && it is Water } && targetHeight == 1) {
            adder.addGraphics(Factory.getGraphics("bang", location, rotation))
        }
    }

    private fun move(deltaTime: Float) {
        val dx = Math.cos(rotation.toDouble()) * speed.toDouble() * deltaTime.toDouble()
        val dy = Math.sin(rotation.toDouble()) * speed.toDouble() * deltaTime.toDouble()
        location = Point(location.x + dx.toFloat(), location.y + dy.toFloat())
        flewDistance += speed * deltaTime
    }

    private fun enemyAt(cell: Cell, enemies: Array<MapObject>, height: Int): Array<MapObject?> {
        if (bang) {
            val e: MutableList<MapObject?> = ArrayList()
            for (`object` in enemies) {
                if (`object` !== owner && `object`.height == height) {
                    val territory = `object`.territory
                    for (c in territory) {
                        if (cell == c) {
                            if (!e.contains(`object`)) e.add(`object`)
                        }
                    }
                    if (`object` is RenderableObject) {
                        val l = (`object` as RenderableObject).location
                        if (l.x - bangRange < location.x && l.x + bangRange > location.x && l.y - bangRange < location.y && l.y + bangRange > location.y) {
                            if (!e.contains(`object`)) e.add(`object`)
                        }
                    }
                }
            }
            return e.toTypedArray()
        } else {
            for (`object` in enemies) {
                if (`object` !== owner && `object`.height == height) {
                    val territory = `object`.territory
                    for (c in territory!!) {
                        if (cell == c) {
                            return arrayOf(`object`)
                        }
                    }
                }
            }
        }
        return arrayOfNulls(0)
    }

    override fun setGoal(goal: Point?) {
        rotation = Math.atan(
                (goal!!.y.toDouble() - location.y.toDouble()) /
                        (goal.x.toDouble() - location.x.toDouble())).toFloat()
        if (goal.x < location.x) {
            rotation += Math.PI.toFloat()
        }
        this.goal = Cell(goal)
    }
}